File: app/src/main/java/de/kuschku/libquassel/CoreConnection.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import de.kuschku.libquassel.client.ClientData;
import de.kuschku.libquassel.client.QClient;
import de.kuschku.libquassel.events.ConnectionChangeEvent;
import de.kuschku.libquassel.events.GeneralErrorEvent;
import de.kuschku.libquassel.events.HandshakeFailedEvent;
import de.kuschku.libquassel.events.UnknownCertificateEvent;
import de.kuschku.libquassel.functions.types.HandshakeFunction;
import de.kuschku.libquassel.functions.types.Heartbeat;
import de.kuschku.libquassel.objects.types.ClientInit;
import de.kuschku.libquassel.primitives.QMetaTypeRegistry;
import de.kuschku.libquassel.primitives.serializers.ProtocolSerializer;
import de.kuschku.libquassel.primitives.types.Protocol;
import de.kuschku.libquassel.protocols.DatastreamPeer;
import de.kuschku.libquassel.protocols.LegacyPeer;
import de.kuschku.libquassel.protocols.RemotePeer;
import de.kuschku.libquassel.ssl.CertificateManager;
import de.kuschku.libquassel.ssl.UnknownCertificateException;
import de.kuschku.util.ServerAddress;
import de.kuschku.util.niohelpers.WrappedChannel;

import static de.kuschku.libquassel.primitives.QMetaType.Type.UInt;
import static de.kuschku.util.AndroidAssert.assertNotNull;

/**
 * Starts a connection to a core and handles the data in the backend.
 * Provides a Client object for interacting with a friendly tree structure of the data.
 *
 * @author Janne Koschinski
 */
public class CoreConnection {

    @NonNull
    private final ServerAddress address;
    @NonNull
    private final ClientData clientData;
    @NonNull
    private final BusProvider busProvider;
    @NonNull
    private final QClient client;
    @NonNull
    private final CertificateManager certificateManager;
    @Nullable
    private ExecutorService outputExecutor;
    @Nullable
    private EndableThread inputThread;
    @Nullable
    private EndableThread heartbeatThread;
    @Nullable
    private RemotePeer remotePeer;
    @Nullable
    private WrappedChannel channel;
    @Nullable
    private Socket socket;
    @NonNull
    private ConnectionChangeEvent.Status status = ConnectionChangeEvent.Status.DISCONNECTED;

    public CoreConnection(@NonNull final ServerAddress address,
                          @NonNull final ClientData clientData,
                          @NonNull final BusProvider busProvider,
                          @NonNull final QClient client,
                          @NonNull CertificateManager certificateManager) {
        this.address = address;
        this.clientData = clientData;
        this.busProvider = busProvider;
        this.client = client;
        this.certificateManager = certificateManager;
    }

    @NonNull
    public ConnectionChangeEvent.Status getStatus() {
        return status;
    }

    /**
     * This method opens a socket to the specified address and starts the connection process.
     *
     * @param supportsKeepAlive If the connection may use keepAlive
     * @throws IOException
     */
    public void open(boolean supportsKeepAlive) throws IOException {
        assertNotNull(client);

        status = ConnectionChangeEvent.Status.HANDSHAKE;
        client.setConnectionStatus(status);

        // Intialize socket
        socket = new Socket();
        if (supportsKeepAlive) socket.setKeepAlive(true);
        socket.connect(new InetSocketAddress(address.host, address.port), 10000);

        // Wrap socket in channel for nio functions
        channel = WrappedChannel.ofSocket(socket);

        busProvider.event.register(this);

        // Create executor for write events
        outputExecutor = Executors.newSingleThreadExecutor();

        // Execute handshake
        handshake();
    }

    /**
     * Closes the connection and interrupts all threads this connection has spawned.
     */
    public void close() {
        assertNotNull(client);

        client.setConnectionStatus(ConnectionChangeEvent.Status.DISCONNECTED);

        // We can do this because we clean up the file handles ourselves
        if (inputThread != null) inputThread.end();
        if (heartbeatThread != null) heartbeatThread.end();
        if (outputExecutor != null) outputExecutor.shutdownNow();

        // Which we do exactly here
        try {
            if (channel != null) channel.close();
            if (socket != null) socket.close();
        } catch (Exception e) {
            // We won’t report these issues, as these don’t matter to us anyway anymore
        }
    }

    @Nullable
    public ExecutorService getOutputExecutor() {
        return outputExecutor;
    }

    @NonNull
    public ClientData getClientData() {
        return clientData;
    }

    @NonNull
    public WrappedChannel getChannel() {
        assertNotNull(channel);

        return channel;
    }

    @Nullable
    public RemotePeer getRemotePeer() {
        return remotePeer;
    }

    @Nullable
    public Socket getSocket() {
        return socket;
    }

    /**
     * Starts the first handshake phase with negotiating the protocol and if SSL or compression are to be used.
     *
     * @throws IOException
     */
    private void handshake() throws IOException {
        assertNotNull(channel);

        // Start protocol handshake with magic version and feature flags
        QMetaTypeRegistry.serialize(UInt, channel, 0x42b33f00 | clientData.flags.flags);

        // Send list of supported protocols
        for (int supportedProtocol : clientData.supportedProtocols) {
            QMetaTypeRegistry.serialize(UInt, channel, supportedProtocol);
        }
        QMetaTypeRegistry.serialize(UInt, channel, 0x01 << 31);

        // Spawn and start a new read thread
        inputThread = new ReadThread();
        heartbeatThread = new HeartbeatThread();
        inputThread.start();
    }

    public void onEventAsync(HandshakeFailedEvent event) {
        this.close();
    }

    public void onEvent(@NonNull ConnectionChangeEvent event) {
        this.status = event.status;
        if (event.status == ConnectionChangeEvent.Status.INITIALIZING_DATA && heartbeatThread != null)
            heartbeatThread.start();
    }

    public void setCompression(boolean supportsCompression) {
        if (supportsCompression)
            channel = WrappedChannel.withCompression(getChannel());
    }

    private void setSSL(boolean supportsSSL) {
        if (supportsSSL) {
            try {
                channel = WrappedChannel.withSSL(getChannel(), certificateManager, address);
            } catch (Exception e) {
                if (e.getCause() instanceof UnknownCertificateException) {
                    busProvider.sendEvent(new UnknownCertificateEvent((UnknownCertificateException) e.getCause()));
                } else {
                    busProvider.sendEvent(new GeneralErrorEvent(e));
                }
                close();
            }
        }
    }

    /**
     * A runnable that reads from the channel and calls the functions responsible for processing the read data.
     */
    private class ReadThread extends EndableThread {
        private boolean running = true;

        public ReadThread() {
            setName(getClass().getSimpleName());
        }

        @Override
        public void run() {
            assertNotNull(client);

            try {
                boolean hasReadPreHandshake = false;
                while (running) {
                    if (!hasReadPreHandshake) {
                        final ByteBuffer buffer = ByteBuffer.allocate(4);
                        assertNotNull(buffer);
                        getChannel().read(buffer);

                        final Protocol protocol = ProtocolSerializer.get().deserialize(buffer);

                        // Wrap socket in SSL context if ssl is enabled
                        setSSL(protocol.protocolFlags.supportsSSL);
                        // Wrap socket in deflater if compression is enabled
                        setCompression(protocol.protocolFlags.supportsCompression);

                        // Initialize remote peer
                        switch (protocol.protocolVersion) {
                            case 0x01:
                                remotePeer = new LegacyPeer(CoreConnection.this, busProvider);
                                break;
                            case 0x02:
                                remotePeer = new DatastreamPeer(CoreConnection.this, busProvider);
                                break;
                            default:
                                busProvider.sendEvent(new HandshakeFailedEvent("Core too new: Protocol Unsupported"));
                                close();
                                return;
                        }

                        // Mark prehandshake as read
                        hasReadPreHandshake = true;
                        assertNotNull(heartbeatThread);

                        // Send client data to core
                        String clientDate = new SimpleDateFormat("MMM dd yyyy HH:mm:ss", Locale.US).format(new Date());
                        assertNotNull(clientDate);
                        busProvider.dispatch(new HandshakeFunction(new ClientInit(
                                clientDate,
                                protocol.protocolFlags.supportsSSL,
                                getClientData().identifier,
                                false,
                                getClientData().protocolVersion
                        )));
                    } else {
                        remotePeer.processMessage();
                    }
                }
            } catch (SocketException e) {
                Log.e("libquassel", "Socket closed while reading");
                client.setConnectionStatus(ConnectionChangeEvent.Status.DISCONNECTED);
            } catch (Exception e) {
                busProvider.sendEvent(new GeneralErrorEvent(e));
            }
        }

        @Override
        public void end() {
            running = false;
        }
    }

    private class HeartbeatThread extends EndableThread {
        private boolean running = true;

        public HeartbeatThread() {
            setName(getClass().getSimpleName());
        }

        @Override
        public void run() {
            try {
                assertNotNull(client);

                while (running) {
                    Heartbeat heartbeat = new Heartbeat();
                    busProvider.dispatch(heartbeat);

                    Thread.sleep(30 * 1000);
                }
            } catch (InterruptedException e) {
            }
        }

        @Override
        public void end() {
            running = false;
        }
    }

    private abstract class EndableThread extends Thread {
        public abstract void end();
    }
}

File: lib/src/main/java/de/kuschku/libquassel/connection/CoreConnection.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2018 Janne Koschinski
 * Copyright (c) 2018 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.connection

import de.kuschku.libquassel.protocol.ClientData
import de.kuschku.libquassel.protocol.message.HandshakeMessage
import de.kuschku.libquassel.protocol.message.SignalProxyMessage
import de.kuschku.libquassel.protocol.primitive.serializer.HandshakeVariantMapSerializer
import de.kuschku.libquassel.protocol.primitive.serializer.IntSerializer
import de.kuschku.libquassel.protocol.primitive.serializer.ProtocolInfoSerializer
import de.kuschku.libquassel.protocol.primitive.serializer.VariantListSerializer
import de.kuschku.libquassel.quassel.ProtocolFeature
import de.kuschku.libquassel.session.ProtocolHandler
import de.kuschku.libquassel.util.Optional
import de.kuschku.libquassel.util.compatibility.CompatibilityUtils
import de.kuschku.libquassel.util.compatibility.HandlerService
import de.kuschku.libquassel.util.compatibility.LoggingHandler.Companion.log
import de.kuschku.libquassel.util.compatibility.LoggingHandler.LogLevel.DEBUG
import de.kuschku.libquassel.util.compatibility.LoggingHandler.LogLevel.WARN
import de.kuschku.libquassel.util.flag.hasFlag
import de.kuschku.libquassel.util.helpers.hexDump
import de.kuschku.libquassel.util.helpers.write
import de.kuschku.libquassel.util.nio.ChainedByteBuffer
import de.kuschku.libquassel.util.nio.WrappedChannel
import io.reactivex.subjects.BehaviorSubject
import java.io.Closeable
import java.lang.Thread.UncaughtExceptionHandler
import java.net.Socket
import java.net.SocketException
import java.nio.ByteBuffer
import javax.net.ssl.SSLSession
import javax.net.ssl.X509TrustManager

class CoreConnection(
  private val clientData: ClientData,
  private val features: Features,
  private val trustManager: X509TrustManager,
  private val hostnameVerifier: HostnameVerifier,
  private val address: SocketAddress,
  private val handlerService: HandlerService
) : Thread(), Closeable {
  companion object {
    private const val TAG = "CoreConnection"
  }

  private var handler: ProtocolHandler? = null
  private var securityExceptionCallback: ((QuasselSecurityException) -> Unit)? = null
  private var exceptionCallback: ((Throwable) -> Unit)? = null

  fun setHandlers(handler: ProtocolHandler?,
                  securityExceptionCallback: ((QuasselSecurityException) -> Unit)?,
                  exceptionCallback: ((Throwable) -> Unit)?) {
    this.handler = handler
    this.securityExceptionCallback = securityExceptionCallback
    this.exceptionCallback = exceptionCallback
  }

  private val exceptionHandler = UncaughtExceptionHandler { thread, throwable ->
    log(WARN, TAG, thread.name, throwable)
  }

  private val sizeBuffer = ByteBuffer.allocateDirect(4)
  private val chainedBuffer = ChainedByteBuffer(direct = true)

  val state: BehaviorSubject<ConnectionState> = BehaviorSubject.createDefault(
    ConnectionState.DISCONNECTED
  )

  private var channel: WrappedChannel? = null
    set(value) {
      field = value
      sslSession.onNext(Optional.ofNullable(value?.sslSession))
    }

  private fun connect() {
    setState(ConnectionState.CONNECTING)
    val socket = Socket()
    if (CompatibilityUtils.supportsKeepAlive)
      socket.keepAlive = true
    socket.connect(address.data(), 10_000)
    handlerService.exceptionHandler = exceptionHandler
    channel = WrappedChannel.ofSocket(socket)
  }

  fun setState(value: ConnectionState) {
    val current = state.value
    if (current != ConnectionState.CLOSED) {
      log(DEBUG, TAG, value.name)
      state.onNext(value)
    } else if (current != value) {
      log(WARN, TAG, "Trying to set state while closed: $value", Throwable())
    }
  }

  private fun sendHandshake() {
    setState(ConnectionState.HANDSHAKE)
    IntSerializer.serialize(
      chainedBuffer,
      0x42b33f00 or clientData.protocolFeatures.toInt(),
      features.negotiated
    )
    for (supportedProtocol in clientData.supportedProtocols) {
      IntSerializer.serialize(chainedBuffer, supportedProtocol.toInt(), features.negotiated)
    }
    IntSerializer.serialize(chainedBuffer, 1 shl 31, features.negotiated)
    channel?.write(chainedBuffer)
    channel?.flush()
  }

  private fun readHandshake() {
    sizeBuffer.clear()
    channel?.read(sizeBuffer)
    sizeBuffer.flip()
    val protocol = ProtocolInfoSerializer.deserialize(sizeBuffer, features.negotiated)

    log(DEBUG, TAG, "Protocol negotiated $protocol")

    // Wrap socket in SSL context if ssl is enabled
    if (protocol.flags.hasFlag(ProtocolFeature.TLS)) {
      channel = channel?.withSSL(trustManager, hostnameVerifier, address)
    }

    // Wrap socket in deflater if compression is enabled
    if (protocol.flags.hasFlag(ProtocolFeature.Compression)) {
      channel = channel?.withCompression()
    }

    // Initialize remote peer
    when (protocol.version.toInt()) {
      0x02 -> {
        // Send client clientData to core
        dispatch(
          HandshakeMessage.ClientInit(
            clientVersion = clientData.identifier,
            buildDate = clientData.buildDate.epochSecond.toString(),
            clientFeatures = clientData.clientFeatures.toInt(),
            featureList = clientData.clientFeatures.toStringList()
          )
        )
      }
      else -> {
        throw ProtocolVersionException(protocol)
      }
    }
  }

  override fun close() {
    try {
      setState(ConnectionState.CLOSED)
      channel?.flush()
      channel?.close()
      setHandlers(null, null, null)
      interrupt()
    } catch (e: Throwable) {
      log(WARN, TAG, "Error encountered while closing connection: $e")
    }
  }

  fun dispatch(message: HandshakeMessage) {
    handlerService.serialize {
      try {
        val data = HandshakeMessage.serialize(message)
        handlerService.write(
          MessageRunnable(
            data, HandshakeVariantMapSerializer, chainedBuffer, channel,
            features.negotiated
          )
        )
      } catch (e: Throwable) {
        log(WARN, TAG, "Error encountered while serializing handshake message", e)
      }
    }
  }

  fun dispatch(message: SignalProxyMessage) {
    handlerService.serialize {
      try {
        val data = SignalProxyMessage.serialize(message)
        handlerService.write(
          MessageRunnable(
            data, VariantListSerializer, chainedBuffer, channel,
            features.negotiated
          )
        )
      } catch (e: Throwable) {
        log(WARN, TAG, "Error encountered while serializing sigproxy message", e)
      }
    }
  }

  override fun run() {
    try {
      connect()
      sendHandshake()
      readHandshake()
      while (!isInterrupted && state != ConnectionState.CLOSED) {
        sizeBuffer.clear()
        if (channel?.read(sizeBuffer) == -1)
          break
        sizeBuffer.flip()

        val size = IntSerializer.deserialize(sizeBuffer, features.negotiated)
        if (size > 64 * 1024 * 1024)
          throw SocketException("Too large frame received: $size")
        val dataBuffer = ByteBuffer.allocateDirect(size)
        while (dataBuffer.position() < dataBuffer.limit() && channel?.read(dataBuffer) ?: -1 > 0) {
        }
        dataBuffer.flip()

        handlerService.deserialize {
          when (state.value) {
            ConnectionState.CLOSED    ->
              // Connection closed, do nothing
              Unit
            ConnectionState.CONNECTING,
            ConnectionState.HANDSHAKE ->
              processHandshake(dataBuffer)
            else                      ->
              processSigProxy(dataBuffer)
          }
        }
      }
      channel?.close()
    } catch (e: Throwable) {
      val closed = state.value == ConnectionState.CLOSED

      var cause: Throwable? = e
      var exception: QuasselSecurityException?
      do {
        exception = cause as? QuasselSecurityException
        cause = cause?.cause
      } while (cause != null && exception == null)
      if (exception != null) {
        close()
        securityExceptionCallback?.invoke(exception)
      } else {
        if (!closed) {
          log(WARN, TAG, "Error encountered in connection", e)
          log(WARN, TAG, "Last sent message: ${MessageRunnable.lastSent.get()}")
          exceptionCallback?.invoke(e)
        }
        close()
      }
    }
  }

  private fun processSigProxy(dataBuffer: ByteBuffer) = handlerService.deserialize {
    try {
      val msg = SignalProxyMessage.deserialize(
        VariantListSerializer.deserialize(dataBuffer, features.negotiated)
      )
      handlerService.backend {
        try {
          handler?.handle(msg)
        } catch (e: Throwable) {
          log(WARN, TAG, "Error encountered while handling sigproxy message", e)
          log(WARN, TAG, msg.toString())
        }
      }

    } catch (e: Throwable) {
      log(WARN,
          TAG, "Error encountered while parsing sigproxy message", e)
      dataBuffer.hexDump()
    }
  }

  private fun processHandshake(dataBuffer: ByteBuffer) = try {
    val msg = HandshakeMessage.deserialize(
      HandshakeVariantMapSerializer.deserialize(dataBuffer, features.negotiated)
    )
    try {
      handler?.handle(msg)
    } catch (e: Throwable) {
      log(WARN,
          TAG, "Error encountered while handling handshake message", e)
      log(WARN, TAG, msg.toString())
    }
  } catch (e: Throwable) {
    log(
      WARN,
      TAG, "Error encountered while parsing handshake message", e
    )
  }


  val sslSession: BehaviorSubject<Optional<SSLSession>> = BehaviorSubject.createDefault(Optional.empty())
}
